plugins/out_gcs/gcs.c

```diff
@@ -0,0 +1,2063 @@
+    char *buffer = NULL;
+    size_t buffer_size = 0;
+    struct mk_list *tmp;
+    struct mk_list *head;
+    int ret;
+    time_t now;
+    int async_flags;
+
+    flb_plg_debug(ctx->ins, "Running upload timer callback (cb_gcs_upload)..");
+
+    /* upload timer must use sync mode */
+    async_flags = flb_stream_get_flags(&ctx->u->base);
+    flb_stream_disable_async_mode(&ctx->u->base);
+
+    now = time(NULL);
+
+    /* Check all chunks and see if any have timed out */
+    mk_list_foreach_safe(head, tmp, &ctx->stream_active->files) {
+        fsf = mk_list_entry(head, struct flb_fstore_file, _head);
+        chunk = fsf->data;
+
+        if (now < (chunk->create_time + ctx->upload_timeout + ctx->retry_time)) {
+            continue; /* Only send chunks which have timed out */
+        }
+
+        /* Locked chunks are being processed, skip */
+        if (chunk->locked == FLB_TRUE) {
+            continue;
+        }
+
+        ret = construct_request_buffer(ctx, NULL, chunk, &buffer, &buffer_size);
+        if (ret < 0) {
+            flb_plg_error(ctx->ins, "Could not construct request buffer for %s",
+                          chunk->file_path);
+            continue;
+        }
+
+        /* FYI: if construct_request_buffer() succeedeed, the gcs_file is locked */
+        ret = upload_data(ctx, chunk, buffer, buffer_size,
+                          (const char *) fsf->meta_buf, fsf->meta_size);
+        flb_free(buffer);
+        if (ret != FLB_OK) {
+            flb_plg_error(ctx->ins, "Could not send chunk with tag %s",
+                          (char *) fsf->meta_buf);
+        }
+    }
+
+    flb_stream_set_flags(&ctx->u->base, async_flags);
+}
+
+static flb_sds_t flb_pack_msgpack_extract_log_key(void *out_context, const char *data,
+                                                  uint64_t bytes)
+{
+    int i;
+    int records = 0;
+    int map_size;
+    int check = FLB_FALSE;
+    int found = FLB_FALSE;
+    int log_key_missing = 0;
+    int ret;
+    int alloc_error = 0;
+    struct flb_gcs *ctx = out_context;
+    char *val_buf;
+    char *key_str = NULL;
+    size_t key_str_size = 0;
+    size_t off = 0;
+    size_t msgpack_size = bytes + bytes / 4;
+    size_t val_offset = 0;
+    flb_sds_t out_buf;
+    msgpack_unpacked result;
+    msgpack_object root;
+    msgpack_object map;
+    msgpack_object key;
+    msgpack_object val;
+
+    /* Iterate the original buffer and perform adjustments */
+    records = flb_mp_count(data, bytes);
+    if (records <= 0) {
+        return NULL;
+    }
+
+    /* Allocate buffer to store log_key contents */
+    val_buf = flb_malloc(msgpack_size);
+    if (val_buf == NULL) {
+        flb_plg_error(ctx->ins, "Could not allocate enough "
+                      "memory to read record");
+        flb_errno();
+        return NULL;
+    }
+
+    msgpack_unpacked_init(&result);
+    while (!alloc_error &&
+           msgpack_unpack_next(&result, data, bytes, &off) == MSGPACK_UNPACK_SUCCESS) {
```

Comment:
We need to use flb_log_event_decoder APIs to decode metadata included events.

* Initialize the decoder: flb_log_event_decoder_init
* Use decoder API to retrive Fluent Bit V2 events: flb_log_event_decoder_next 
* Teardown the decoder: flb_log_event_decoder_destroy